src/lib/rule2F.ts

```diff
@@ -138,6 +138,93 @@ function allowsNameFromContent(elem: HTMLElement): boolean {
 
 export const TEST_ONLY = {allowsNameFromContent};
 
+// See https://lists.w3.org/Archives/Public/public-aria/2017Jun/0057.html
+// for discussion of roles & tags that forbid name from content.
+const NO_NAME_FROM_CONTENT_ROLES = [
+  'application',
+  'alert',
+  'log',
+  'marquee',
+  'timer',
+  'alertdialog',
+  'dialog',
+  'banner',
+  'complementary',
+  'form',
+  'main',
+  'navigation',
+  'region',
+  'search',
+  'article',
+  'document',
+  'feed',
+  'figure',
+  'img',
+  'math',
+  'toolbar',
+  'menu',
+  'menubar',
+  'grid',
+  'listbox',
+  'radiogroup',
+  'textbox',
+  'searchbox',
+  'spinbutton',
+  'scrollbar',
+  'slider',
+  'tablist',
+  'tabpanel',
+  'tree',
+  'treegrid',
+  'separator',
+  'rowgroup',
+  'group',
+];
+const NO_NAME_FROM_CONTENT_TAGS = [
+  'article',
+  'aside',
+  'body',
+  'select',
+  'datalist',
+  'optgroup',
+  'dialog',
+  'figure',
+  'footer',
+  'form',
+  'header',
+  'hr',
+  'img',
+  'textarea',
+  'input',
+  'main',
+  'math',
+  'menu',
+  'nav',
+  'section',
+  'thead',
+  'tbody',
+  'tfoot',
+  'fieldset',
+];
+
+/**
+ * Checks if 'elem' is forbidden from allowing 'name from content'
+ * @param elem - element whose text alternative is being computed
+ */
+function forbidsNameFromContent(elem: HTMLElement): boolean {
```

Comment:
From the names I'd expect `allowsNameFromContent(el) === !forbidsNameFromContent(el);`. How about combining this logic into `allowsNameFromContent`?

Reply:
I think the naming might be misleading here -- I could possibly change the wording to 'roleAllowsNameFromContent' / 'roleForbidsNameFromContent' to clarify.

The idea here being that in 2F we might use the content (subtree) of the current node as part of the name computation in the following cases:
- its role allows name from content
- it is part of a label
- it is descendent of a node whose accessible name is being computed

But if the role forbids name from content, then we shouldn't look at the subtree even if the current node is part of a label etc.

I could combine these into a single function `allowsNameFromContent` that looks at current node role, whether it's part of a label etc. all in the one function? This could even be called `rule2FCondition`.